File: 0703-kth-largest-element-in-a-stream/0703-kth-largest-element-in-a-stream.py
import heapq

class KthLargest:
    def __init__(self, k: int, nums: List[int]):
        self.num_elems = k
        self.pq_elems = []
        for num in nums:
            heapq.heappush(self.pq_elems, num)
            if len(self.pq_elems) > k:
                heapq.heappop(self.pq_elems)

    def add(self, val: int) -> int:
        if len(self.pq_elems) < self.num_elems:
            heapq.heappush(self.pq_elems, val)
        else:
            if val > self.pq_elems[0]:
                heapq.heappop(self.pq_elems)
                heapq.heappush(self.pq_elems, val)
        return self.pq_elems[0]

File: 0703-kth-largest-element-in-a-stream/0703-kth-largest-element-in-a-stream.cpp
class KthLargest {
private:
    priority_queue <int, vector<int>, greater<int>> pq_elems;
    int num_elems;
public:
    KthLargest(int k, vector<int>& nums) {
        num_elems = k;
      
        for(int i=0; i<nums.size(); i++)
        {
            pq_elems.push(nums[i]);
            if(pq_elems.size()>k)
                pq_elems.pop();
            
        }
    }
    
    int add(int val) {
        if(pq_elems.size() < num_elems)
            pq_elems.push(val);
         else
         {
             if(val > pq_elems.top())
            {
                pq_elems.pop();
                pq_elems.push(val);
            }
         }
        return pq_elems.top();
        
    }
};
